Add methods to add custom reference BEM data, and fix UWG code-style.

saeranv opened the pull request:
@mostaphaRoudsari, @chriswmackey 

Thanks for being patient with this PR. It's still in draft form, but the major parts of the new UWG object are in place, including:

- exposed all UWG macro-scale properties with class getters and setters (see at [uwg/uwg.py](https://github.com/ladybug-tools/uwg/pull/127/files?file-filters%5B%5D=.pkl&file-filters%5B%5D=.py&file-filters%5B%5D=.txt&file-filters%5B%5D=.uwg&file-filters%5B%5D=dotfile#diff-a226d4381c852ce81325ea5b7180fdc2))
- added from_dict and to_dict methods corresponding to the uwg-schema for UWG, and associated building, construction and schedules objects 
- code changed to PEP8 style, and adding docstrings. Thus far 12/18 modules, and all the tests have had their style fixed (see caveat below re: variable case).
- Added methods (i.e `customize_reference_data`) to the UWG which will permit the modification of individual buildings, constructions, and schedules.

Things that still need to be completed:
- add getters/setters for the exposed building, construction and schedule objects
- most of the variable names haven't been converted to (PEP8-consistent) lowercase yet, 
- finish PEP8 style revision of remaining 6 modules 

I'll let you know when this is ready for a full code-review,

chriswmackey commented:
@saeranv ,
Thanks for starting this and I'm sorry that it took so long for me to notice it.  I will review first thing tomorrow,

chriswmackey commented on uwg/material.py:
```diff
@@ -1,22 +1,78 @@
 """Material class"""
 
+from honeybee.typing import float_in_range_excl_incl
```
Again here.

chriswmackey commented on uwg/uwg.py:
```diff
@@ -13,6 +13,7 @@
 
 from __future__ import division, print_function
 from functools import reduce
+from honeybee.typing import int_in_range, float_in_range, int_positive, float_positive
```
And here.

chriswmackey reviewed (commented):
Hey @saeranv ,
This is definitely an improvement and I just made a few comments on the styling. I also think you should remove the `honeybee.typing` dependency sooner than later.  I trust your judgement here with whatever you can get done in terms of styling.  I realize it's a lot that has to be changed and I think it's totally ok if you PEP8-ify 50% of it, get the tests to pass, and the rest can be taken care of later.

saeranv commented on uwg/RSMDef.py:
```diff
@@ -6,59 +7,89 @@
+        self.z_meso = RSMDef.load_z_meso(z_meso_path)
+        self.lat = lat  # latitude [deg]
+        self.lon = lon  # longitude [deg]
+        self.GMT = GMT  # GMT hour correction
```
Thanks, changed it. There isn't really a strategy for which variables I'm changing and which ones I'm not. I'm changing all the variables that will be exposed to the user (i.e inputs that are specified in the uwg-schema), but for the rest, the only bottleneck is just quantity of variables. Almost everything is in camelcase right now, defined across multiple modules, tests which is tricky to resolve. PyCharm has the ability to refactor variables across multiple modules, but sometimes it misses some, (and it isn't so good with tests) so every variable change takes a bit of effort!

saeranv commented on uwg/RSMDef.py:
```diff
@@ -74,314 +105,323 @@ def __init__(self,lat,lon,GMT,height,T_init,P_init,parameter,z_meso_path):
-            if (eq_th == True or self.z[iz] > parameter.tempHeight) and ll==True:
-                self.nz0 = iz+1   # layer number at zmt (m)
+            eq_th = is_near_zero(self.z[iz] - parameter.tempHeight)
+            if (eq_th is True or self.z[iz] > parameter.tempHeight) and ll is True:
```
Yeah I know... I just never really liked this PEP8 recommendation because I think it causes confusion when you combine it with Python's dynamic typing.  

For example if greeting wasn't a boolean. 
```
greeting = 1
if greeting: print('x')
>> x
```
And there's equivalent versions for lists, None and other types. So my instinct is to explicitly define what's being tested in an if/else condition so I don't have to hunt around looking for what the variable type is when rereading/editing code. 

BUT, it's not a big deal, so I'll change it to the implicit version.

chriswmackey replied:
Maybe it also helps to know that PEP8 only recommends using `if x` and `if not x` in the case that `x` is a boolean.  You are not supposed to use it in the case that it's a None (for that you should be using `if x is None`).  And you also shouldn't use it for the case that `x` is a number.  So you always know the type is a boolean if you're following PEP8.

saeranv replied:
@chriswmackey ah! Finally that makes sense to me.

saeranv commented on uwg/RSMDef.py:
```diff
@@ -74,314 +105,323 @@ def __init__(self,lat,lon,GMT,height,T_init,P_init,parameter,z_meso_path):
-    # Ref: The uwg (2012), Eq. (4)
-    def VDM(self,forc,rural,parameter,simTime):
+    def VDM(self, forc, rural, parameter, simTime):
+        """# Ref: The uwg (2012), Eq. (4)"""
```
Good catch. I've tried to do this everywhere but clearly missed this function. I'll keep an eye out, but let me know if I missed defining the abbreviations anywhere else.

saeranv commented on uwg/element.py:
```diff
@@ -6,172 +7,281 @@
 import math
+from .utilities import is_near_zero
+from .material import Material
+from honeybee.typing import float_in_range, float_positive
```
Done.

saeranv commented on uwg/RSMDef.py:
```diff
@@ -74,314 +105,323 @@ def __init__(self,lat,lon,GMT,height,T_init,P_init,parameter,z_meso_path):
-
-    def DiffusionCoefficient(self,rho,z,dz,z0,disp,tempRur,heatRur,nz,uref,th,parameter):
+    def DiffusionCoefficient(self, rho, z, dz, z0, disp, tempRur, heatRur, nz, uref, th,
+                             parameter):
```
@chriswmackey 

I can change the capitalization (and I'll hunt down other methods where function capitilization occurs), but adding docstrings are a bit of a challenge. 

This is the one module in UWG that I have very little understanding of because it's relying on an understanding of physics that's beyond me (mesoscale fluid dynamics). Additionally, the original Matlab code uses nondescriptive, single-letter variable naming for almost everything, which is really confusing. I'd have to dig really deep into Bruno's thesis to be able to define docstrings. It is okay to keep it as is for now?

chriswmackey replied:
Yes, don't worry about it if you don't actually know what the variables are.  Bruno should have been the one to document it and, now that he didn't, tracking down what it means is a whole other big task unto itself.  For now, it might help to put a `# TODO:` here that at least let's other folk know that writing the docstring requires us to do some mesoscale fluid dynamic research.

saeranv commented on uwg/delete_me_scratch:
```diff
@@ -0,0 +1,21 @@
+--utilities.py
+--uwg.py
+--weather.py
+
```
Sorry didn't meant to commit this, it's just some record keeping, will delete.

saeranv commented:
@chriswmackey 

Here's the latest draft. I addressed all the comments (replaced honeybee.typing, style changes, minor code revision) you brought up in the PR feedback, but am still chipping away at PEP8 conversions in some of the other modules. 

This new draft includes the CLI commands for UWG. Specifically I added two commands to simulate the UWG, and validate files. Both simulation and validation work for the model JSON and .uwg parameter file. 

One question I had was, does the UWG CLI need to log it's output with logger? I noticed you have it in honeybee-energy, but UWG doesn't produce copious error logs like E+, so just printed everything to stdout for UWG. Let me know if I should add it in.

saeranv commented:
@chriswmackey 

Just note that all the changes I made to the [tree/vegetation heat calculations](https://github.com/ladybug-tools/uwg/issues/107) broke some of my old unit tests. In the process of fixing them.

saeranv commented:
@chriswmackey FYI, 

All tests are working now, but travis is failing the cli tests since it's requires the latest uwg-schema PR to be pushed. Once that's done we'll be good.

chriswmackey commented:
Thank @saeranv . This is looking great. I'll review the schema PR now and, if it looks good, I will merge.

saeranv commented:
@chriswmackey  

I just finished adding the changes we discussed from uwg-schema to this PR. Should be good to go.

saeranv commented on uwg/uwg.py:
```diff
@@ -886,133 +1323,550 @@ def simulate(self):
+                self.BEM[i].building.coolcap = 9999.
+                self.BEM[i].building.heat_cap = 9999.
+
+    def _check_reference_data(self, ref_bem_vector, ref_sch_vector):
```
Check reference data

saeranv commented:
@chriswmackey 

This PR now includes the changes to the way we define reference BEMs and Schedules in bld and BEMDef/SchDef. From now on bld consists of a list of tuples, with each tuple defining bldtype, builtera, and fraction of built stock:
```
     bld = [('largeoffice', 'pst80', 0.4),
                ('midriseapartment', 'new', 0.3),
                ('custombuilding', 'pre80', 0.3)]
```
The bldtype, and builtera property in BEMDef/SchDef also are defined with these strings. Note the .frac property in BEMDef is now a private property, so the only way to define the fraction of urban stock of a reference building is with the bld list. 

Similarily I also modified the UWG.zone so that it's a string and not based on index in zone, i.e:

```
Get or set text representing an ASHRAE climate zone.

        Choose from the following:

        * '1A' - (i.e Miami)
        * '2A' - (i.e Houston)
        * '2B' - (i.e Phoenix)
        * '3A' - (i.e Atlanta)
        * '3B-CA' - (i.e Los Angeles)
        * '3B' - (i.e Las Vegas)
        * '3C' - (i.e San Francisco)
        * '4A' - (i.e Baltimore)
        * '4B' - (i.e Albuquerque)
        * '4C' - (i.e Seattle)
        * '5A' - (i.e Chicago)
        * '5B' - (i.e Boulder)
        * '6A' - (i.e Minneapolis)
        * '6B' - (i.e Helena)
        * '7' - (i.e Duluth)
        * '8' - (i.e Fairbanks)
```

saeranv commented:
@chriswmackey 

Just pushed some changes to keep this consistent with changes we're making in the uwg-schema.

saeranv commented:
@chriswmackey 

Latest PR just keeps things consistent with the defaults just added to the uwg-schema. 

Also, I want to double-check one of my assumptions regarding the failing test. Travis is failing specifically because it can't find and install the uwg-schema model (the test_cli.py works on my local). I've added the following line in my setup.py so that the uwg-schema will be downloaded.
```
extras_require={
        'cli': ['click==7.1.2', "uwg-schema;python_version>='3.6'"]
    },
```

Am I correct in assuming that once the pip version of the uwg-schema is released the failing test will resolve itself? Or do I need to do something else?
